### PR TITLE
added probe implementation to CompositeDiscoveryClient

### PR DESCRIPTION
**Pull request by sruffatti**
Fixes gh-1209

### CODE REVIEW

**Review by spencergibb — CHANGES_REQUESTED** [review]
Please add a test

**Comment by sruffatti** [comment]
Any additional feedback on this @spencergibb ?

**Comment by sruffatti** [comment]
Any updates on this @spencergibb ?

**Comment by sruffatti** [comment]
bump @spencergibb

**Comment by sruffatti** [comment]
@OlgaMaciaszek Are you able to help me out here? This PR has been open for a while.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+     @Mock
+     private DiscoveryClient client2;
+ 
+     @BeforeEach
+     void setUp() {
+         underTest = new CompositeDiscoveryClient(Arrays.asList(client1, client2));
+     }
+ 
+     @Test
+     void getDescriptionShouldBeComposite() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java:59]: Please remove this test - we don't really need to test hard-coded `String` values.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+         then(underTest.description()).isEqualTo("Composite Discovery Client");
+     }
+ 
+     @Test
+     void getInstancesByServiceIdShouldReturnServiceInstances() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java:64]: Please rename the test to `shouldRetrieveInstancesByServiceId`.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+         List<ServiceInstance> serviceInstances = underTest.getInstances("serviceId");
+ 
+         then(serviceInstances.get(0).getInstanceId()).isEqualTo("instance1");
+         then(serviceInstances.get(0).getServiceId()).isEqualTo("serviceId");
+         then(serviceInstances.get(0).getHost()).isEqualTo("https://s1");
+         then(serviceInstances.get(0).getPort()).isEqualTo(8443);
+     }
+ 
+     @Test
+     void getServicesShouldReturnServiceIds() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java:77]: Please rename the test to `shouldReturnServiceIds`.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+         when(client2.getServices()).thenReturn(Collections.singletonList("serviceId2"));
+ 
+         List<String> services = underTest.getServices();
+ 
+         then(services.size()).isEqualTo(2);
+         then(services).containsOnlyOnce("serviceId1", "serviceId2");
+     }
+ 
+     @Test
+     void getDiscoveryClientsShouldReturnAllDiscoveryClients() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java:88]: Please rename the test to `shouldReturnAllDiscoveryClients`.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+         then(underTest.getDiscoveryClients()).containsOnlyOnce(client1, client2);
+     }
+ 
+     @Test
+     void probeShouldCallProbeInAllClientBeans() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java:93]: Please rename to `shouldCallProbeOnAllDiscoveryClients`.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2012-2020 the original author or authors.
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClientUnitTests.java:2]: Please update the copyright dates to `2012-2023`.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClient.java [hunk 7]
```diff
@@ -73,6 +73,15 @@ public List<String> getServices() {
 		return new ArrayList<>(services);
 	}
 
+	@Override
```
Comment (OlgaMaciaszek) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/composite/CompositeDiscoveryClient.java:76]: Please add your name with `@author` tag in the javadoc of the edited file. Please update the copyright date line.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hi @sruffatti, thanks for submitting the PR. Makes sense. There are some minor style issues. I have added comments - please address them. Also, could you please file this against `3.1.x` instead of `main`? It'll be considered a bugfix, so we'd like for it to also be included in the `2021.0.x` release train.

**Comment by sruffatti** [comment]
> Hi @sruffatti, thanks for submitting the PR. Makes sense. There are some minor style issues. I have added comments - please address them. Also, could you please file this against `3.1.x` instead of `main`? It'll be considered a bugfix, so we'd like for it to also be included in the `2021.0.x` release train.

@OlgaMaciaszek I addressed your comments. Are you asking me to change the target branch to `3.1.x` instead `main`? When I do that I see more commits. Are you asking me to branch off `3.1.x`, add my changes, and then open a PR?

**Comment by OlgaMaciaszek** [comment]
Hi @sruffatti,
> @OlgaMaciaszek I addressed your comments. Are you asking me to change the target branch to 3.1.x instead main? When I do that I see many commits included in the PR. Are you asking me to branch off 3.1.x, add my changes, and then open a PR? 

The second approach will probably be faster and easier, so it's a good idea.

**Comment by sruffatti** [comment]
@OlgaMaciaszek I created another PR merging to 3.1.x here https://github.com/spring-cloud/spring-cloud-commons/pull/1239

**Comment by OlgaMaciaszek** [comment]
Closing in favour of https://github.com/spring-cloud/spring-cloud-commons/pull/1239.

**Pull request closed without merging**